### PR TITLE
validator public keys listing fix

### LINKED ISSUE #6976: Accounts-V2: listing accounts produces incorrect output mixing up "validating public key" entries
# 🐞 Bug Report

### Description

When listing V2 validator accounts, the output is incorrect, the "validating public key" entries are mixed up, showing as belonging to different accounts than they really do.

An example can be seen in ##6918, copied below:

<pre>
  <code>
Account 0 | constantly-popular-wasp
[validating public key] 0xb692725fe769d79863eefe4752fac00f878072b8c898a8fbfca32189692ceb0bdb77212d82d9ee8a199b48a60461263b

Account 1 | cheaply-summary-gull
[validating public key] 0x8329d8e9aa7da5f6b8300a22e55fd5657d6fa42d30408c83218da10136db06936e1bd7025574718e4283a1adf6c1ab71

Account 2 | carefully-wondrous-terrapin
[validating public key] 0x99cc6a756b0492b2790903a175351d1e153b513e1cfb95db7a56a33cd811c5f7c859d94cea46b67b8e7e6481f09eb5f5

Account 3 | repeatedly-prime-stinkbug
[validating public key] 0xa9e8ab052e4e80670ecf54a3bec7f43a08a1b146063221348a5c735e42ef16db1c0c9685f29192502526f91272b80a96
  </code>
</pre>

<pre>
  <code>
Account 0 | constantly-popular-wasp
[withdrawal public key] 0xa387211380b13ab47d59a00f3eaa7a6edffd1f1b9745231aa607eb7480c42e403dba11d00a8e4f5d7c80627677c3c050
[derivation path] m/12381/3600/0/0
[validating public key] 0xa9e8ab052e4e80670ecf54a3bec7f43a08a1b146063221348a5c735e42ef16db1c0c9685f29192502526f91272b80a96
[derivation path] m/12381/3600/0/0/0

Account 1 | cheaply-summary-gull
[withdrawal public key] 0xb8ac604b685fd322c21ce1e2f263934c7e272cb6d3e60052430d6975d2666c2f142ba042213101c48c333df95a40abf4
[derivation path] m/12381/3600/1/0
[validating public key] 0x9958265fbd2577270ad8fda3745bd804ff77fb50fe02b781f203cd571204178e60686f4052803f7ceb5334ac197c2dd8
[derivation path] m/12381/3600/1/0/0

Account 2 | carefully-wondrous-terrapin
[withdrawal public key] 0xaf0f953920a8203b1d0a96f59e1cbfb80961086f10931c2ddaa8250d4d952dc38482847da15816842de56417f8273ab5
[derivation path] m/12381/3600/2/0
[validating public key] 0xb692725fe769d79863eefe4752fac00f878072b8c898a8fbfca32189692ceb0bdb77212d82d9ee8a199b48a60461263b
[derivation path] m/12381/3600/2/0/0

Account 3 | repeatedly-prime-stinkbug
[withdrawal public key] 0x87b9ebcd8af91ad43485aa311888f1622a2c41ce1614c24d205e9f52506c28c0543628c3eff8db18845925236421fc91
[derivation path] m/12381/3600/3/0
[validating public key] 0x8329d8e9aa7da5f6b8300a22e55fd5657d6fa42d30408c83218da10136db06936e1bd7025574718e4283a1adf6c1ab71
[derivation path] m/12381/3600/3/0/0
  </code>
</pre>

For example the key `0xb692725fe769d79863eefe4752fac00f878072b8c898a8fbfca32189692ceb0bdb77212d82d9ee8a199b48a60461263b` is in the first listing show as belonging to the account `constantly-popular-wasp`, in the second listing it is shown to belong to the account `carefully-wondrous-terrapin`.

### Has this worked before in a previous version?

No.

## 🔬 Minimal Reproduction

Generate multiple accounts, note the validator public keys, then invoke validator again listing the accounts and observe that the `[validating public key]` entries are mixed up.

## 🔥 Error

No error.

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
Distribution:      Clear Linux OS
Installed version: 33620

Docker version 19.03.8, build afacb8b7f0d8d4f9d2a8e8736e9c993e672b41f3
  </code>
</pre>

**What version of Prysm are you running? (Which release)**

<pre>
  <code>
gcr.io/prysmaticlabs/prysm/validator:v1.0.0-alpha.19
  </code>
</pre>

**Anything else relevant (validator index / public key)?**

The methods producing the output are calling the respective `keymanager.FetchValidatingPublicKeys` functions to retrieve the list of validator public keys. The direct and derived keymanagers both use the key collection of a hashmap to in the `FetchValidatingPublicKeys` function to produce the lis. Because of how a hashmap is implemented, that list is produced in a different order than the order in which the accounts are stored and listed, resulting in a mixup.

Here is the relevant code:

https://github.com/prysmaticlabs/prysm/blob/399dc85737639df62f3dc2f7d95478307812440f/validator/keymanager/v2/direct/direct.go#L311-L318

https://github.com/prysmaticlabs/prysm/blob/399dc85737639df62f3dc2f7d95478307812440f/validator/keymanager/v2/derived/derived.go#L346-L353

### PR DESCRIPTION
**Pull request by RoBiK75**
**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**

Account v2 list for direct and derived accounts produced wrong output, mixing up validator public keys.

**Which issues(s) does this PR fix?**

Fixes #6976

**Other notes for review**

### CODE REVIEW

File: validator/keymanager/v2/derived/derived.go [hunk 1]
```diff
@@ -258,6 +258,20 @@ func (dr *Keymanager) ValidatingAccountNames(ctx context.Context) ([]string, err
 	return names, nil
 }
 
+// ValidatingPublicKeys for the derived keymanager.
+func (dr *Keymanager) ValidatingPublicKeys() ([][48]byte, error) {
```
Comment (terencechain) [inline validator/keymanager/v2/derived/derived.go:262]: Is this related to the fix? I have a feeling we might need to split this into two prs

Comment (RoBiK75) [reply validator/keymanager/v2/derived/derived.go:262]: Yes, this is part of the fix. The only thing that is not directly a fix of this problem but is related is this:
https://github.com/prysmaticlabs/prysm/pull/6978/commits/18ff2fb6522854c6c126622b6e124cee59572162

Comment (terencechain) [reply validator/keymanager/v2/derived/derived.go:262]: Thanks for confirming. I think that's ok to include. 
Perhaps add a note in the description and mention there's 2 fixes and the intention of each fix

**Comment by RoBiK75** [comment]
After discussion with @rauljordan this PR is being closed in favor of an another solution.
